gms/feature_service.cc

```diff
@@ -248,10 +248,20 @@ future<> feature_service::enable_features_on_join(gossiper& g, db::system_keyspa
     return enabler->enable_features();
 }
 
-future<> feature_service::enable_features_on_startup(db::system_keyspace& sys_ks) {
+future<> feature_service::enable_features_on_startup(db::system_keyspace& sys_ks, bool use_raft_cluster_features) {
     std::set<sstring> features_to_enable;
-    const auto persisted_features = co_await sys_ks.load_local_enabled_features();
-    if (persisted_features.empty()) {
+    std::set<sstring> persisted_features;
+    std::set<sstring> persisted_unsafe_to_disable_features;
+
+    if (!use_raft_cluster_features) {
+        persisted_features = co_await sys_ks.load_local_enabled_features();
+    } else {
+        auto topo = co_await sys_ks.load_topology_state();
+        persisted_features = std::move(topo.enabled_features);
+        persisted_unsafe_to_disable_features = topo.calculate_not_yet_enabled_features();
```

Comment:
This might not work after the `move` above

Reply:
Cursed non-destructive move semantics...!

It works, but by accident, because `persisted_unsafe_to_disable_features` will include both "not yet enabled features" and the enabled features from `persisted_features`. I'll fix it, of course.